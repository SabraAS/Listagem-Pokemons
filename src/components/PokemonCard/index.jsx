import PropTypes from 'prop-types';

import coverPokemon from '@/assets/cover-pokemon.png';
import './index.scss';

const PokemonCard = ({
  abilities,
  addPokemon,
  characteristic,
  disabled,
  id,
  image,
  name,
  types,
}) => {
  const getValues = (values) =>
    values?.length
      ? values
          .map((item) => item?.type?.name || item?.ability?.name)
          .filter(Boolean)
          .join(', ')
      : 'não possui';

  return (
    <div className="pokemon-card" data-testid="pokemon-card">
      <img
        alt={`Imagem do ${name || 'pokémon'}`}
        className="pokemon-card__image"
        data-testid="pokemon-card-image"
        src={image || coverPokemon}
      />
      <button
        aria-label={
          disabled
            ? `${name || 'pokémon'} indisponível`
            : `Adicionar ${name || 'pokémon'} à equipe`
        }
        className="pokemon-card__button"
        data-testid={`pokemon-card-button-${id}`}
        disabled={disabled}
        onClick={() => addPokemon(id || name)}
      >
        {disabled ? 'Indisponível' : 'Adicionar à equipe'}
      </button>
      <h2 className="pokemon-card__name" data-testid="pokemon-card-name">
        {name || 'pokémon sem nome'}
      </h2>
      <div className="pokemon-card__info">
        <p
          className="pokemon-card__text"
          data-testid="pokemon-card-characteristic"
        >
          Característica: {characteristic || 'não possui'}
        </p>
        <p className="pokemon-card__text" data-testid="pokemon-card-abilities">
          Habilidades: {getValues(abilities)}
        </p>
        <p className="pokemon-card__text" data-testid="pokemon-card-types">
          Tipos: {getValues(types)}
        </p>
      </div>
    </div>
  );
};

PokemonCard.propTypes = {
  addPokemon: PropTypes.func.isRequired,
  abilities: PropTypes.array,
  characteristic: PropTypes.string,
  disabled: PropTypes.bool,
  id: PropTypes.number,
  image: PropTypes.string,
  name: PropTypes.string,
  types: PropTypes.array,
};

export default PokemonCard;
