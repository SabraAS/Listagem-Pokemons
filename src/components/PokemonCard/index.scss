.pokemon-card {
  align-items: center;
  border-radius: 0.5rem;
  display: flex;
  flex-direction: column;
  font-family: Poppins, sans-serif;
  justify-content: flex-start;
  overflow: hidden;
  width: 250px;

  &__image {
    border-radius: 16px;
    height: 250px;
    object-fit: contain;
    width: 100%;
  }

  &__button {
    background: white;
    border: 2px solid #f4dc26;
    border-radius: 25px;
    box-shadow: 0 2px 4px rgb(0 0 0 / 10%);
    color: #374151;
    cursor: pointer;
    font-family: Poppins, sans-serif;
    font-size: 14px;
    font-weight: 500;
    margin: 4px 0 16px;
    padding: 12px 0;
    text-align: center;
    transition: all 0.2s ease-in-out;
    width: 160px;

    &:hover:not(:disabled) {
      box-shadow: 0 4px 8px rgb(0 0 0 / 10%);
      transform: scale(1.05);
    }

    &:disabled {
      background: #e5e7eb;
      border-color: #e5e7eb;
      color: #9ca3af;
      cursor: not-allowed;
    }
  }

  &__content {
    text-align: left;
  }

  &__name {
    color: #1f2937;
    font-size: 1.3rem;
    font-weight: 600;
    margin: 8px 0;
  }

  &__info {
    color: #4b5563;
    display: block;
    font-size: 0.8rem;
    margin-bottom: 4px;
  }
}
