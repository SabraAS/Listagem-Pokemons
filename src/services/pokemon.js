import axios from 'axios';

const BASE_URL = 'https://pokeapi.co/api/v2';

export const getPokemonList = async (limit = 40) => {
  try {
    const response = await axios.get(`${BASE_URL}/pokemon?limit=${limit}`);
    const pokemonList = response.data.results;

    return await Promise.all(
      pokemonList.map(async (pokemon) => {
        const pokemonData = await axios.get(pokemon.url);
        const description = await getPokemonCharacteristic(pokemonData.data.id);
        return {
          id: pokemonData.data.id,
          name: pokemonData.data.name,
          image: pokemonData.data.sprites.front_default,
          abilities: pokemonData.data.abilities,
          types: pokemonData.data.types,
          characteristic: description,
        };
      }),
    );
  } catch (error) {
    console.error('Erro ao buscar lista de Pokémon:', error);
    throw error;
  }
};

export const getPokemonCharacteristic = async (id) => {
  try {
    const response = await fetch(
      `https://pokeapi.co/api/v2/characteristic/${id}`,
    );
    const data = await response.json();
    const englishDescription = data.descriptions.find(
      (desc) => desc.language.name === 'en',
    );
    return englishDescription?.description || '';
  } catch (e) {
    console.error('Erro ao buscar característica do Pokémon:', e);
    return '';
  }
};
