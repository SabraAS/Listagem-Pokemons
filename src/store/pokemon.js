import { create } from 'zustand';

export const usePokemonStore = create((set) => ({
  pokemons: [],
  setPokemons: (pokemons) => set({ pokemons }),
  addPokemon: (pokemon) =>
    set((state) => ({ pokemons: [...state.pokemons, pokemon] })),
  removePokemon: (pokemonId) =>
    set((state) => ({
      pokemons: state.pokemons.filter((pokemon) => pokemon.id !== pokemonId),
    })),
  clearTeam: () => set({ pokemons: [] }),
}));
