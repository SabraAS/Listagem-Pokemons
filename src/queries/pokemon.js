import { useQuery } from '@tanstack/react-query';

import { getPokemonList } from '@/services/pokemon';

export const POKEMON_KEYS = {
  all: ['pokemons'],
  list: (limit) => [...POKEMON_KEYS.all, 'list', limit],
};

export const usePokemons = (limit = 20) => {
  return useQuery({
    queryKey: POKEMON_KEYS.list(limit),
    queryFn: () => getPokemonList(limit),
    retry: false,
    refetchOnWindowFocus: false,
    select: (response) => response,
  });
};
