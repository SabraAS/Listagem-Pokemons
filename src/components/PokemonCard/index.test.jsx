import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';

import PokemonCard from './index';

import coverPokemon from '@/assets/cover-pokemon.png';
import { mockSinglePokemon } from '@/test/mocks/pokemon';

describe('PokemonCard', () => {
  const defaultProps = {
    id: mockSinglePokemon.id,
    name: mockSinglePokemon.name,
    image: mockSinglePokemon.image,
    characteristic: mockSinglePokemon.characteristic,
    abilities: mockSinglePokemon.abilities,
    types: mockSinglePokemon.types,
    disabled: false,
    addPokemon: vi.fn(),
  };

  describe('Rendering', () => {
    it('should match snapshot', () => {
      const { container } = render(<PokemonCard {...defaultProps} />);
      expect(container).toMatchSnapshot();
    });

    it('should render pokemon name correctly', () => {
      render(<PokemonCard {...defaultProps} />);
      const name = screen.getByTestId('pokemon-card-name');
      expect(name).toHaveTextContent('bulbasaur');
    });

    it('should render pokemon image with correct alt text', () => {
      render(<PokemonCard {...defaultProps} />);
      const image = screen.getByTestId('pokemon-card-image');
      expect(image).toHaveAttribute('src', mockSinglePokemon.image);
    });

    it('should render characteristic correctly', () => {
      render(<PokemonCard {...defaultProps} />);
      const characteristic = screen.getByTestId('pokemon-card-characteristic');
      expect(characteristic).toHaveTextContent(
        'Característica: A strange seed was planted on its back at birth. The plant sprouts and grows with this Pokémon.',
      );
    });

    it('should render abilities correctly', () => {
      render(<PokemonCard {...defaultProps} />);
      const abilities = screen.getByTestId('pokemon-card-abilities');
      expect(abilities).toHaveTextContent('Habilidades: overgrow, chlorophyll');
    });

    it('should render types correctly', () => {
      render(<PokemonCard {...defaultProps} />);
      const types = screen.getByTestId('pokemon-card-types');
      expect(types).toHaveTextContent('Tipos: grass, poison');
    });

    it('should render add button when not disabled', () => {
      render(<PokemonCard {...defaultProps} />);
      const button = screen.getByTestId('pokemon-card-button-1');
      expect(button).toHaveTextContent('Adicionar à equipe');
      expect(button).not.toBeDisabled();
    });

    it('should render disabled button when disabled', () => {
      render(<PokemonCard {...defaultProps} disabled={true} />);
      const button = screen.getByTestId('pokemon-card-button-1');
      expect(button).toHaveTextContent('Indisponível');
      expect(button).toBeDisabled();
    });
  });

  describe('Interactions', () => {
    it('should call addPokemon when button is clicked', () => {
      const mockAddPokemon = vi.fn();
      render(<PokemonCard {...defaultProps} addPokemon={mockAddPokemon} />);
      const button = screen.getByTestId('pokemon-card-button-1');
      fireEvent.click(button);
      expect(mockAddPokemon).toHaveBeenCalledTimes(1);
      expect(mockAddPokemon).toHaveBeenCalledWith(defaultProps.id);
    });

    it('should not call addPokemon when disabled button is clicked', () => {
      const mockAddPokemon = vi.fn();
      render(
        <PokemonCard
          {...defaultProps}
          addPokemon={mockAddPokemon}
          disabled={true}
        />,
      );
      const button = screen.getByTestId('pokemon-card-button-1');
      fireEvent.click(button);
      expect(mockAddPokemon).not.toHaveBeenCalled();
    });

    it('should pass correct id on each call', () => {
      const mockAddPokemon = vi.fn();
      render(<PokemonCard {...defaultProps} addPokemon={mockAddPokemon} />);
      const button = screen.getByTestId('pokemon-card-button-1');
      fireEvent.click(button);
      expect(mockAddPokemon).toHaveBeenCalledWith(1); // id do bulbasauro
    });
  });

  describe('Edge cases', () => {
    it('should handle empty characteristic', () => {
      render(<PokemonCard {...defaultProps} characteristic="" />);
      const characteristic = screen.getByTestId('pokemon-card-characteristic');
      expect(characteristic).toHaveTextContent('Característica: não possui');
    });

    it('should handle empty name', () => {
      render(<PokemonCard {...defaultProps} name={''} />);
      const name = screen.getByTestId('pokemon-card-name');
      expect(name).toHaveTextContent('pokémon sem nome');

      const image = screen.getByTestId('pokemon-card-image');
      expect(image).toHaveAttribute('src', mockSinglePokemon.image);
      expect(image).toHaveAttribute('alt', 'Imagem do pokémon');

      const button = screen.getByTestId('pokemon-card-button-1');
      expect(button).toHaveTextContent('Adicionar à equipe');
    });

    it('should handle null image', () => {
      render(<PokemonCard {...defaultProps} image={null} />);
      const image = screen.getByTestId('pokemon-card-image');
      expect(image).toHaveAttribute('src', coverPokemon);
      expect(image).toHaveAttribute('alt', 'Imagem do bulbasaur');
    });

    it('should handle null abilities and types', () => {
      render(<PokemonCard {...defaultProps} abilities={null} types={null} />);
      const abilities = screen.getByTestId('pokemon-card-abilities');
      expect(abilities).toHaveTextContent('Habilidades: não possui');

      const types = screen.getByTestId('pokemon-card-types');
      expect(types).toHaveTextContent('Tipos: não possui');
    });

    it('should handle null id by using name instead', () => {
      const mockAddPokemon = vi.fn();
      render(
        <PokemonCard {...defaultProps} addPokemon={mockAddPokemon} id={null} />,
      );
      const button = screen.getByTestId('pokemon-card-button-null');
      fireEvent.click(button);
      expect(mockAddPokemon).toHaveBeenCalledWith('bulbasaur');
    });
  });

  describe('CSS structure', () => {
    it('should have correct main container class', () => {
      render(<PokemonCard {...defaultProps} />);
      const container = screen.getByText('bulbasaur').closest('.pokemon-card');
      expect(container).toBeInTheDocument();
      expect(container).toHaveClass('pokemon-card');
    });

    it('should have correct image class', () => {
      render(<PokemonCard {...defaultProps} />);
      const image = screen.getByAltText('Imagem do bulbasaur');
      expect(image).toHaveClass('pokemon-card__image');
    });

    it('should have correct name class', () => {
      render(<PokemonCard {...defaultProps} />);
      const name = screen.getByText('bulbasaur');
      expect(name).toHaveClass('pokemon-card__name');
    });

    it('should have correct info container class', () => {
      render(<PokemonCard {...defaultProps} />);
      const infoContainer = screen
        .getByText(
          'Característica: A strange seed was planted on its back at birth. The plant sprouts and grows with this Pokémon.',
        )
        .closest('.pokemon-card__info');
      expect(infoContainer).toBeInTheDocument();
      expect(infoContainer).toHaveClass('pokemon-card__info');
    });

    it('should have correct button class', () => {
      render(<PokemonCard {...defaultProps} />);
      const button = screen.getByRole('button');
      expect(button).toHaveClass('pokemon-card__button');
    });

    it('should have correct text classes for all text elements', () => {
      render(<PokemonCard {...defaultProps} />);
      const characteristicText = screen.getByText(
        'Característica: A strange seed was planted on its back at birth. The plant sprouts and grows with this Pokémon.',
      );
      const abilitiesText = screen.getByText(
        'Habilidades: overgrow, chlorophyll',
      );
      const typesText = screen.getByText('Tipos: grass, poison');
      expect(characteristicText).toHaveClass('pokemon-card__text');
      expect(abilitiesText).toHaveClass('pokemon-card__text');
      expect(typesText).toHaveClass('pokemon-card__text');
    });

    it('should apply disabled styles to button when disabled', () => {
      render(<PokemonCard {...defaultProps} disabled={true} />);
      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
      expect(button).toHaveClass('pokemon-card__button');
      expect(button).toHaveTextContent('Indisponível');
    });
  });

  describe('Accessibility', () => {
    it('should have no accessibility violations', async () => {
      const { container } = render(<PokemonCard {...defaultProps} />);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('should have no accessibility violations when disabled', async () => {
      const { container } = render(
        <PokemonCard {...defaultProps} disabled={true} />,
      );
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('should have proper button accessibility', () => {
      render(<PokemonCard {...defaultProps} />);
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute(
        'aria-label',
        'Adicionar bulbasaur à equipe',
      );
    });

    it('should have proper button accessibility when disabled', () => {
      render(<PokemonCard {...defaultProps} disabled={true} />);
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('aria-label', 'bulbasaur indisponível');
    });

    it('should have proper button accessibility with null name', () => {
      render(<PokemonCard {...defaultProps} name={null} />);
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute(
        'aria-label',
        'Adicionar pokémon à equipe',
      );
    });

    it('should have proper button accessibility when disabled with null name', () => {
      render(<PokemonCard {...defaultProps} disabled={true} name={null} />);
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('aria-label', 'pokémon indisponível');
    });

    it('should have proper button accessibility with undefined name', () => {
      render(<PokemonCard {...defaultProps} name={undefined} />);
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute(
        'aria-label',
        'Adicionar pokémon à equipe',
      );
    });

    it('should have proper button accessibility when disabled with undefined name', () => {
      render(
        <PokemonCard {...defaultProps} disabled={true} name={undefined} />,
      );
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('aria-label', 'pokémon indisponível');
    });

    it('should have proper image accessibility', () => {
      render(<PokemonCard {...defaultProps} />);
      const image = screen.getByRole('img');
      expect(image).toHaveAttribute('alt', 'Imagem do bulbasaur');
    });

    it('should have proper heading structure', () => {
      render(<PokemonCard {...defaultProps} />);
      const heading = screen.getByText('bulbasaur');
      expect(heading.tagName).toBe('H2');
    });

    it('should be keyboard accessible', () => {
      render(<PokemonCard {...defaultProps} />);
      const button = screen.getByRole('button');
      button.focus();
      expect(document.activeElement).toBe(button);
    });

    it('should have proper focus management when disabled', () => {
      render(<PokemonCard {...defaultProps} disabled={true} />);
      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    it('should have descriptive text content', () => {
      render(<PokemonCard {...defaultProps} />);
      expect(screen.getByText(/Característica:/)).toBeInTheDocument();
      expect(screen.getByText(/Habilidades:/)).toBeInTheDocument();
      expect(screen.getByText(/Tipos:/)).toBeInTheDocument();
    });
  });
});
