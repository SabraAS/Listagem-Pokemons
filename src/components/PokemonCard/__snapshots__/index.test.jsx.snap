// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokemonCard > Rendering > should match snapshot 1`] = `
<div>
  <div
    class="pokemon-card"
    data-testid="pokemon-card"
  >
    <img
      alt="Imagem do bulbasaur"
      class="pokemon-card__image"
      data-testid="pokemon-card-image"
      src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg"
    />
    <button
      aria-label="Adicionar bulbasaur à equipe"
      class="pokemon-card__button"
      data-testid="pokemon-card-button-1"
    >
      Adicionar à equipe
    </button>
    <h2
      class="pokemon-card__name"
      data-testid="pokemon-card-name"
    >
      bulbasaur
    </h2>
    <div
      class="pokemon-card__info"
    >
      <p
        class="pokemon-card__text"
        data-testid="pokemon-card-characteristic"
      >
        Característica: 
        A strange seed was planted on its back at birth. The plant sprouts and grows with this Pokémon.
      </p>
      <p
        class="pokemon-card__text"
        data-testid="pokemon-card-abilities"
      >
        Habilidades: 
        overgrow, chlorophyll
      </p>
      <p
        class="pokemon-card__text"
        data-testid="pokemon-card-types"
      >
        Tipos: 
        grass, poison
      </p>
    </div>
  </div>
</div>
`;
